pkg/directory/farms_handlers.go

```diff
@@ -147,46 +128,121 @@ func (f *FarmAPI) getFarm(r *http.Request) (interface{}, mw.Response) {
 }
 
 func (f *FarmAPI) deleteNodeFromFarm(r *http.Request) (interface{}, mw.Response) {
-	sid := mux.Vars(r)["farm_id"]
+	farm := r.Context().Value("farm").(directory.Farm)
 
-	id, err := strconv.ParseInt(sid, 10, 64)
-	if err != nil {
-		return nil, mw.BadRequest(err)
-	}
+	var nodeAPI NodeAPI
+	nodeID := mux.Vars(r)["node_id"]
 
 	db := mw.Database(r)
-
-	farm, err := f.GetByID(r.Context(), db, id)
+	node, err := nodeAPI.Get(r.Context(), db, nodeID, false)
 	if err != nil {
 		return nil, mw.NotFound(err)
 	}
 
-	sfarmerID := httpsig.KeyIDFromContext(r.Context())
-	requestFarmerID, err := strconv.ParseInt(sfarmerID, 10, 64)
+	if node.FarmId != int64(farm.ID) {
+		return nil, mw.Forbidden(fmt.Errorf("only the farm owner of this node can delete this node"))
+	}
+
+	err = nodeAPI.Delete(r.Context(), db, nodeID)
 	if err != nil {
-		return nil, mw.BadRequest(err)
+		return nil, mw.Error(err)
 	}
 
-	if farm.ThreebotId != requestFarmerID {
-		return nil, mw.Forbidden(fmt.Errorf("only the farm owner of this node can delete this node"))
+	return nil, mw.Ok()
+}
+
+func (f *FarmAPI) addFarmIPs(r *http.Request) (interface{}, mw.Response) {
+	// Get the farm from the middleware context
+	farm := r.Context().Value("farm").(directory.Farm)
+
+	var info []generated.FarmerIP
+	if err := json.NewDecoder(r.Body).Decode(&info); err != nil {
+		return nil, mw.BadRequest(err)
 	}
 
-	var nodeAPI NodeAPI
-	nodeID := mux.Vars(r)["node_id"]
+	for _, ip := range info {
+		for _, farmerIP := range farm.IPs {
+			if ip.IP.Equal(farmerIP.IP) {
+				return nil, mw.BadRequest(fmt.Errorf("ip %v already exists in farm", ip))
+			}
+			// If IP does not exist yet, add it to the list
+			farm.IPs = append(farm.IPs, farmerIP)
+		}
+	}
 
-	node, err := nodeAPI.Get(r.Context(), db, nodeID, false)
+	db := mw.Database(r)
+	err := f.Update(r.Context(), db, farm.ID, farm)
 	if err != nil {
-		return nil, mw.NotFound(err)
+		return nil, mw.Error(err)
 	}
 
-	if node.FarmId != int64(farm.ID) {
-		return nil, mw.Forbidden(fmt.Errorf("only the farm owner of this node can delete this node"))
+	return nil, mw.Ok()
+}
+
+func (f *FarmAPI) deleteFarmIps(r *http.Request) (interface{}, mw.Response) {
+	// Get the farm from the middleware context
+	farm := r.Context().Value("farm").(directory.Farm)
+
+	var info []generated.FarmerIP
+	if err := json.NewDecoder(r.Body).Decode(&info); err != nil {
+		return nil, mw.BadRequest(err)
 	}
 
-	err = nodeAPI.Delete(r.Context(), db, nodeID)
+	for _, ip := range info {
+		for idx, farmIP := range farm.IPs {
+			if ip.Equal(farmIP.IP) {
+				if farmIP.ReservationID != 0 {
+					return nil, mw.BadRequest(fmt.Errorf("cannot remove an IP that is in use"))
+				}
+				// If IP is not used, remove it from the farmer's ip's
+				farm.IPs = append(farm.IPs[:idx], farm.IPs[idx+1:]...)
+			}
+		}
+	}
+
+	db := mw.Database(r)
+	err := f.Update(r.Context(), db, farm.ID, farm)
```

Comment:
Again add proper methods that do add/remove ip from mongo arrays properly. Avoid re-saving the entire farm object. mongo has the operation to insert-to-set (array) or remove atomically